### PR TITLE
Improve unit bubbling for ISO dates

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## What is this?

This PR intends to fix issues with invalid ISO dates getting created from a well formatted `moment.duration`. GitHub Issues I am aware of:

* https://github.com/moment/moment/issues/2162
* https://github.com/moment/moment/issues/1697

What I have done is basically:

* not use `this.data` as source of truth anymore for `toISOString`
    * I thought about changing the way `data` is filled, but I believe this would be the wrong approach, correct me if I am wrong
* use some custom bubbling for `toISOString` that stops bubbling at **hours** and **days**
    * these are basically the borders that are also used internally
    * internal fields are milliseconds (which can bubble up to hours), days and month (becoming years)
* add one-line spec that should cover all new edge cases

## Why is it useful?

The ISO 8601 standard (which I only know from second hand knowledge), makes a difference between hours, days and months. That is:

* `P1D` means tommorrow, at the same time
* `PT24H` means in 24 hours, which can be different from the former case, if the clock changed in the meantime (think of DST changes, leap seconds, etc)
* `P30D` means in thirty days, no matter how many days the current month has
* `P1M` means on the same date in the next month

Thus from an ISO perspective there is a difference between `moment.duration(1, 'days')` and `moment.duration(24, 'hours')` which should now be reflected.

*Note: I've written down similar arguments in the linked issues, but I believe it helps clarity to repeat it if I open a PR ^^*

### CODE REVIEW

**Comment by NobodysNightmare** [comment]
Seems like the current `develop` branch is unstable. Travis reported two failures that I can reproduce locally on the `develop` branch alone.

I rebased my PR onto master, so that travis should report it green. Local results:

````
┌───────┬───────┬────────────┬────────┬────────┬─────────┐
│ Files │ Tests │ Assertions │ Failed │ Passed │ Runtime │
├───────┼───────┼────────────┼────────┼────────┼─────────┤
│ 1     │ 2377  │ 45713      │ 0      │ 45713  │ 6596    │
└───────┴───────┴────────────┴────────┴────────┴─────────┘

````

File: src/lib/duration/iso-string.js [hunk 1]
```diff
@@ -1,13 +1,37 @@
...
+
+
     // inspired by https://github.com/dordille/moment-isoduration/blob/master/moment.isoduration.js
-    var Y = abs(this.years());
-    var M = abs(this.months());
-    var D = abs(this.days());
-    var h = abs(this.hours());
-    var m = abs(this.minutes());
-    var s = abs(this.seconds() + this.milliseconds() / 1000);
+    var Y = years;
```
Comment (myabc) [inline src/lib/duration/iso-string.js]: @NobodysNightmare you should be able to use destructuring assignment here:

```js
var [Y, M, D, h, m, s] = [year, months, days, hours, minutes, seconds];
```

Comment (NobodysNightmare) [inline src/lib/duration/iso-string.js]: @myabc good idea, but:

> We're using ES6 module system, but nothing else ES6

Comment (myabc) [inline src/lib/duration/iso-string.js]: @NobodysNightmare ah :(

**Comment by ichernev** [comment]
@NobodysNightmare thank you for working on this!

> not use this.data as source of truth anymore for `toISOString`

You did the right thing. We store milliseconds and days separately exactly for this reason (days should go over DST as it doesn't exist, and hours, minutes, seconds should advance time exactly)

> Seems like the current develop branch is unstable.

Sorry for that. Your change looks good. When I merge I make sure all tests pass (on top of the other changes scheduled for the release).

Will merge in next release.

**Comment by ichernev** [comment]
Merged in 6349dbf8279c5d86be2b18deac5a465dbf7f49c3

**Pull request closed without merging**